File: Grind75/75-LargestRectangleInHistorgram/cpp/LargestRectangleInHistorgram.cpp
#include "LargestRectangleInHistorgram.h"
#include <cstdio>
#include <stack>
#include <sys/types.h>
#include <vector>

using namespace std;

int Solution::largestRectangleArea(vector<int>& heights) {
  heights.push_back(-1);
  int maxArea = 0;
  stack<int> startStack, heightStack;
  for (uint end = 0; end < heights.size(); end++) {
    int start = end, height = heights[start];
    while (!startStack.empty() && heights[end] < heightStack.top()) {
      start = startStack.top();
      height = heightStack.top();
      int width = end - start;
      int area = height * width;
      if (area > maxArea)
        maxArea = area;
      startStack.pop();
      heightStack.pop();
    }
    startStack.push(start);
    heightStack.push(heights[end]);
  }
  return maxArea;
}

File: Grind75/75-LargestRectangleInHistorgram/python/LargestRectangleInHistorgram.py
#!/bin/python3

from __future__ import annotations
import sys
from typing import Tuple
sys.path.append('../../Utility/python')
from PrettyPrint import *
from Colors import *

class Solution:
  def largestRectangleArea(self, heights: List[int]) -> int:
    stack: List[Tuple[int, int]] = []
    maxArea = 0
    heights.append(-1)
    for end, height in enumerate(heights):
      start = end
      while stack and height < stack[-1][1]:
        start, h = stack.pop()
        area = h * (end - start)
        if area > maxArea:
          maxArea = area
      stack.append((start, height))
    return maxArea

if __name__ == '__main__':
  soln = Solution()
  testCases = [
    ( [], 0 ),
    ( [1], 1 ),
    ( [1,2], 2 ),
    ( [2,1], 2 ),
    ( [2,1], 2 ),
    ( [2,1,2], 3 ),
    ( [2,1,4], 4 ),
    ( [2,1,0,2], 2 ),
    ( [2,1,4,3,1], 6 ),
    ( [2,1,4,3,1,1,1], 7 ),
    ( [2,1,2,2,2], 6 ),
    ( [3,3,3,1,3,3], 9 ),
    ( [3,1,5,6,3,4,2], 12 ),
    ( [3,6,5,7,4,8,1,0], 20 ),
  ]

  for heights, answer in testCases:
    colorPrint(LIGHT_BLUE, heights, '=>', answer)
    result = soln.largestRectangleArea(heights)
    colorPrint(GREEN if result == answer else RED, '-->', result)

